### PR TITLE
refactor RocketMQProduceFactory: reuse producer

### PR DESCRIPTION
**Pull request by LeBW**
### Describe what this PR does / why we need it
Refactor RocketMQProduceFactory: reuse producer

### Describe how you did it
1. add a Map `private static Map<String, DefaultMQProducer> producers = new HashMap<>();` to store producers in the factory.
2. reuse the producer in the factory

### Describe how to verify it
Create several output bindings, and the producer will be reused.

### CODE REVIEW

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java [hunk 1]
```diff
@@ -63,7 +68,10 @@ public static DefaultMQProducer initRocketMQProducer(String topic,
 				"Property 'group' is required - producerGroup");
 		Assert.notNull(producerProperties.getNameServer(),
 				"Property 'nameServer' is required");
-
+		String key = getKey(producerProperties);
+		if (producers.containsKey(key)) {
+			return producers.get(key);
```
Comment (Sorieee) [inline spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java]: 缩进问题

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java [hunk 2]
```diff
@@ -45,12 +49,13 @@
  */
 public final class RocketMQProduceFactory {
 
+	private final static Logger log = LoggerFactory.getLogger(RocketMQProduceFactory.class);
+
+	private static Map<String, DefaultMQProducer> producers = new HashMap<>();
```
Comment (Sorieee) [inline spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java]: new HashMap<>() 必须设置容量。
详情参考 阿里巴巴的开发手册。

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java [hunk 3]
```diff
@@ -134,4 +144,34 @@ public static DefaultMQProducer initRocketMQProducer(String topic,
...
+		public void start() throws MQClientException {
+			if (atomicInteger.getAndIncrement() == 0) {
+				super.start();
+			}
+		}
+
+		@Override
+		public void shutdown() {
+			if (atomicInteger.decrementAndGet() == 0) {
+				producers.remove(key);
```
Comment (Sorieee) [inline spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java]: 可能存在并发风险。

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java [hunk 4]
```diff
@@ -134,4 +144,34 @@ public static DefaultMQProducer initRocketMQProducer(String topic,
 		return producer;
 	}
 
+	private static String getKey(RocketMQProducerProperties producerProperties) {
+		return producerProperties.getNameServer() + "," + producerProperties.getGroup();
+	}
+
+	protected static class ReusableMQProducer extends DefaultMQProducer {
```
Comment (Sorieee) [inline spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java:167]: 所有代码都没有注释

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java [hunk 5]
```diff
@@ -63,7 +68,10 @@ public static DefaultMQProducer initRocketMQProducer(String topic,
 				"Property 'group' is required - producerGroup");
 		Assert.notNull(producerProperties.getNameServer(),
 				"Property 'nameServer' is required");
-
+		String key = getKey(producerProperties);
+		if (producers.containsKey(key)) {
```
Comment (Sorieee) [inline spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java]: 并发风险

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java [hunk 6]
```diff
@@ -134,4 +144,34 @@ public static DefaultMQProducer initRocketMQProducer(String topic,
...
+		private final String key;
+
+		public ReusableMQProducer(String namespace, String group, RPCHook rpcHook, boolean enableMsgTrace, String customizedTraceTopic, String key) {
+			super(namespace, group, rpcHook, enableMsgTrace, customizedTraceTopic);
+			this.key = key;
+		}
+
+		@Override
+		public void start() throws MQClientException {
+			if (atomicInteger.getAndIncrement() == 0) {
```
Comment (Sorieee) [inline spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java:181]: 并发风险 是否存在正在启动的时候，其他线程 使用正在启动的producer。

**Review by Sorieee — COMMENTED** [review]
存在一些代码规范、并发风险

**Comment by Sorieee** [comment]
Maybe Spring cloud stream had cache producer, we don't need to cache producer in spring cloud alibaba.
lQLPJxaSQP3GCDjNA4rNBROw6pTMheW3XzUC8LBK44DSAA_1299_906

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java [hunk 7]
```diff
@@ -45,12 +49,14 @@
  */
 public final class RocketMQProduceFactory {
 
-	private RocketMQProduceFactory() {
-	}
-
 	private final static Logger log = LoggerFactory
 			.getLogger(RocketMQProduceFactory.class);
 
+	private static Map<String, DefaultMQProducer> PRODUCER_REUSABLE_MAP = new ConcurrentHashMap<>();
```
Comment (steverao) [inline spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java]: Please add `final` to  PRODUCER_REUSABLE_MAP.

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java (hunk 1 above)
Comment (LeBW) [reply spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java]: fixed

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java (hunk 2 above)
Comment (LeBW) [reply spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java]: 似乎并没有强制设置容量？

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java (hunk 4 above)
Comment (LeBW) [reply spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java:167]: fixed

File: spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java (hunk 7 above)
Comment (LeBW) [reply spring-cloud-alibaba-starters/spring-cloud-starter-stream-rocketmq/src/main/java/com/alibaba/cloud/stream/binder/rocketmq/integration/outbound/RocketMQProduceFactory.java]: fixed

**Comment by LeBW** [comment]
@steverao @zkzlx @yuhuangbin This PR may be reviewed and merged

**Review by steverao — APPROVED** [review]

**Review by yuhuangbin — APPROVED** [review]

**Comment by steverao** [comment]
Don't forget to create a same PR to branch of 2021.x.

**Pull request merged**